grpc/grpcmetrics/highcardmetrics/server.go

```diff
@@ -0,0 +1,162 @@
+		return handler(srv, wrapped)
+	}
+}
+
+// serverStream provides a light wrapper over grpc.ServerStream
+// to instrument SendMsg and RecvMsg.
+type serverStream struct {
+	reg metricsregistry.Registry
+	grpc.ServerStream
+	labels []string
+}
+
+//// RecvMsg implements the grpc.Stream interface.
+func (ss *serverStream) SendMsg(m interface{}) (err error) {
+	defer func(begin time.Time) {
+		instrumentStreamSend(ss.reg, ss.labels, time.Since(begin), err)
+	}(time.Now())
+
+	return ss.ServerStream.SendMsg(m)
+}
+
+//// RecvMsg implements the grpc.Stream interface.
+func (ss *serverStream) RecvMsg(m interface{}) (err error) {
+	defer func(begin time.Time) {
+		instrumentStreamRecv(ss.reg, ss.labels, time.Since(begin), err)
+	}(time.Now())
+
+	return ss.ServerStream.RecvMsg(m)
+}
+
+func instrumentMethod(r metricsregistry.Registry, labels []string, duration time.Duration) {
+	r.GetOrRegisterHistogram("request-duration.ms", 50).With(labels...).Observe(ms(duration))
+	r.GetOrRegisterCounter("requests").With(labels...).Add(1)
+}
+
+func instrumentStreamSend(r metricsregistry.Registry, labels []string, duration time.Duration, err error) {
+	r.GetOrRegisterHistogram("stream.send-duration.ms", 50).With(labels...).Observe(ms(duration))
+	r.GetOrRegisterCounter("stream.sends").With(labels...).Add(1)
+
+	if err != nil && !isCanceled(err) {
+		r.GetOrRegisterCounter("stream.sends.errors").Add(1)
+	}
+}
+
+func instrumentStreamRecv(r metricsregistry.Registry, labels []string, duration time.Duration, err error) {
+	r.GetOrRegisterHistogram("stream.recv-duration.ms", 50).With(labels...).Observe(ms(duration))
+	r.GetOrRegisterCounter("stream.recvs").With(labels...).Add(1)
+
+	if err != nil && !isCanceled(err) {
+		r.GetOrRegisterCounter("stream.recvs.errors").With(labels...).Add(1)
+	}
+}
+
+func parseFullMethod(fullMethod string) (string, string) {
+	parts := strings.Split(fullMethod, "/")
+	if len(parts) < 3 {
+		return grpcmetrics.Unknown, grpcmetrics.Unknown
+	}
+
+	fullService := parts[1]
+	method := parts[2]
+
+	sp := strings.Split(fullService, ".")
+	service := sp[len(sp)-1]
+
+	return dasherize(service), dasherize(method)
+}
+
+// code returns the gRPC error code, handling context and unknown errors.
+func code(err error) string {
+	if err == context.Canceled {
+		return grpcmetrics.Canceled
+	}
+
+	st, ok := status.FromError(err)
+	if !ok {
+		return grpcmetrics.Unknown
+	}
+
+	return dasherize(st.Code().String())
+}
+
+var uppers = regexp.MustCompile(`([[:lower:]])([[:upper:]])`)
+
+func dasherize(s string) string {
+	return strings.ToLower(uppers.ReplaceAllString(s, "$1-$2"))
+}
+
+// isCanceled returns true if error is a context or gRPC cancelation error.
+func isCanceled(err error) bool {
+	if err == context.Canceled {
+		return true
+	}
+
+	if st, ok := status.FromError(err); ok {
+		return st.Code() == codes.Canceled
+	}
+
+	return false
+}
+
+func ms(d time.Duration) float64 {
+	return float64(d) / float64(time.Millisecond)
+}
```

Comment:
There is a much easier way to do this now https://pkg.go.dev/time#Duration.Milliseconds